src/Serilog/Parameters/PropertyBinder.cs

```diff
@@ -65,31 +65,33 @@ IEnumerable<LogEventProperty> ConstructPositionalProperties(MessageTemplate temp
                 SelfLog.WriteLine("Positional property count does not match parameter count: {0}", template);
 
             var result = new LogEventProperty[messageTemplateParameters.Length];
-            foreach (var property in positionalProperties)
+            int nextUnmatchedPosition = 0;
+            for (int position = 0; position < messageTemplateParameters.Length; position++)
             {
-                int position;
-                if (property.TryGetPositionalValue(out position))
+                PropertyToken propertyToken = null;
+                for (int propIndex = 0; propIndex < positionalProperties.Length; propIndex++)
                 {
-                    if (position < 0 || position >= messageTemplateParameters.Length)
-                        SelfLog.WriteLine("Unassigned positional value {0} in: {1}", position, template);
-                    else
-                        result[position] = ConstructProperty(property, messageTemplateParameters[position]);
+                    int propertyTokenPosition;
+                    if (positionalProperties[propIndex].TryGetPositionalValue(out propertyTokenPosition)
+                        && propertyTokenPosition == position)
+                    {
+                        propertyToken = positionalProperties[propIndex];
+                    }
                 }
-            }
 
-            var next = 0;
-            for (var i = 0; i < result.Length; ++i)
-            {
-                if (result[i] != null)
+                if (propertyToken == null)
                 {
-                    result[next] = result[i];
-                    ++next;
+                    // found no property for this position; generate a default name
+                    result[position] = new LogEventProperty(
+                        "__" + nextUnmatchedPosition++,
```

Comment:
What do you think about using `i` for this, so that for example in `("{A}", a, b)` the property generated for `b` will be named `__1`?

When the example is `("{0}", a, b)` this would make the two properties `0` and `__1`, rather than `0` and `__0` as in the code above.

It feels natural to me to consider the underscored names to be based on the index of the parameter rather than its index in the list of unmatched properties, but curious if there are advantages to the alternative.